frontend/kubernetes/plugin.go

```diff
@@ -1508,3 +1543,191 @@ func (p *Plugin) updatePVCSize(
+}
+
+func (p *Plugin) processVolumeSnapshot(volSnap *k8ssnapshotv1alpha1.VolumeSnapshot) {
+
+	// Validate the VolumeSnapshot CR
+	// 1. Error must not be set
+	if volSnap.Status.Error != nil {
+		return
+	}
+
+	// 2. Snapshot must not be created before
+	if volSnap.Status.CreationTime != nil && volSnap.Status.ReadyToUse {
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+		}).Debug("Kubernetes frontend ignored this VolumeSnapshot as " +
+			"the snapshot has already been created")
+		return
+	}
+
+	// 3. Snapshot source must be of PVC kind
+	if volSnap.Spec.Source.Kind != "PersistentVolumeClaim" {
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+		}).Debug("Kubernetes frontend ignored this VolumeSnapshot as " +
+			"the source is not a PVC")
+		return
+	}
+
+	log.WithFields(log.Fields{
+		"volumeSnapshot": volSnap.Name,
+		"namespace":      volSnap.Namespace,
+		"sourceKind":     volSnap.Spec.Source.Kind,
+		"sourceName":     volSnap.Spec.Source.Name,
+	}).Debug("Kubernetes frontend got notified of a VolumeSnapshot.")
+
+	// Validate the source PVC
+	// 1. PVC and VolumeSnapshotCR must be in the same namespace
+	sourcePVC := volSnap.Spec.Source.Name
+	k8sClient, err := p.getNamespacedKubeClient(&p.kubeConfig, volSnap.Namespace)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"namespace": volSnap.Namespace,
+			"error":     err.Error(),
+		}).Error("Kubernetes frontend couldn't create a client to namespace")
+		return
+	}
+	claim, err := k8sClient.GetPVC(sourcePVC, metav1.GetOptions{})
+	if err != nil {
+		err = fmt.Errorf("snapshotting a PVC requires the PVC to be in the same namespace as the VolumeSnapshot")
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+			"sourcePVC":      sourcePVC,
+		}).Errorf("Kubernetes frontend detected an invalid configuration "+
+			"for creating snapshot from a PVC: %v", err.Error())
+		return
+	}
+
+	// 2. Trident should be set as the provisioner for the PVC
+	provisioner := getClaimProvisioner(claim)
+	if provisioner != AnnOrchestrator &&
+		GetPersistentVolumeClaimClass(claim) != "" {
+		log.WithFields(log.Fields{
+			"volumeSnapshot":    volSnap.Name,
+			"sourcePVC":         claim.Name,
+			"sourceProvisioner": provisioner,
+		}).Debugf("Kubernetes frontend ignored this VolumeSnapshot as the source PVC is not "+
+			"tagged with %s as the storage provisioner!", AnnOrchestrator)
+		return
+	}
+
+	// 3. PVC must be Bound to a PV
+	if claim.Status.Phase != v1.ClaimBound {
+		log.WithFields(log.Fields{
+			"volumeSnapshot":  volSnap.Name,
+			"sourcePVC":       claim.Name,
+			"sourcePVCStatus": claim.Status.Phase,
+		}).Debug("Kubernetes frontend ignored this VolumeSnapshot as the source PVC is not " +
+			"bound to any volume")
+		return
+	}
+
+	// Create the volume configuration object
+	uniqueClaimName := getUniqueClaimName(claim)
+
+	// Create the snapshot
+	snapshotExt, err := p.orchestrator.CreateVolumeSnapshot(volSnap.Name, uniqueClaimName)
```

Comment:
I must be missing something fundamental here.  volSnap is a VolumeSnapshot object, which is a namespaced, user-visible object, but we are persisting it here in Trident (which manages objects across all namespaces), implying that volSnap.Name is unique in the cluster.  Is that true?

And where is the code that creates VolumeSnapshotContent objects?  My expectation is that Trident would notice a new VolumeSnapshot and react by creating a new VolumeSnapshotContent object (whose name Trident would create and could guarantee would be unique in the cluster).  That would be consistent with how Trident notices new PVCs and creates associated PVs.

Reply:
Hi, @pavannd1, did you have any thoughts on my comments here?  I'm not sure we have a workable approach here in Part 2, which is why I've held off merging the Part 1 patch.

Reply:
@clintonk Sorry for the delayed response. Yes, these are some good points and I have been trying out a couple different solutions for this. I will get back to you with the details about my tests soon.